server/src/main/java/org/elasticsearch/indices/SystemIndices.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.indices;
+
+import org.apache.lucene.util.automaton.Automata;
+import org.apache.lucene.util.automaton.Automaton;
+import org.apache.lucene.util.automaton.CharacterRunAutomaton;
+import org.apache.lucene.util.automaton.MinimizationOperations;
+import org.apache.lucene.util.automaton.Operations;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.regex.Regex;
+import org.elasticsearch.index.Index;
+
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.toUnmodifiableList;
+
+public class SystemIndices {
+
+    private final CharacterRunAutomaton runAutomaton;
+    private final Collection<SystemIndexDescriptor> systemIndexDescriptors;
+
+    public SystemIndices(Map<String, Collection<SystemIndexDescriptor>> systemIndexDescriptorMap) {
+        this.systemIndexDescriptors = systemIndexDescriptorMap.values()
+            .stream()
+            .flatMap(Collection::stream)
+            .collect(Collectors.toUnmodifiableList());
+        this.runAutomaton = buildCharacterRunAutomaton(systemIndexDescriptors);
+    }
+
+    public boolean isSystemIndex(Index index) {
```

Comment:
Javadoc:
```
Determines whether a given index is a system index by comparing its name to the collection of loaded SystemIndexDescriptors.
```

I think this one is good to call out, as most other places this is determined do it by inspecting a property of the index.